modules/wallet/scan.go

```diff
@@ -0,0 +1,198 @@
+package wallet
+
+import (
+	"fmt"
+
+	"github.com/NebulousLabs/Sia/build"
+	"github.com/NebulousLabs/Sia/modules"
+	"github.com/NebulousLabs/Sia/types"
+)
+
+// numInitialKeys is the number of keys generated by the seedScanner before
+// scanning the blockchain for the first time.
+var numInitialKeys = func() uint64 {
+	switch build.Release {
+	case "dev":
+		return 1e4
+	case "standard":
+		return 1e6
+	case "testing":
+		return 1e3
+	default:
+		panic("unrecognized build.Release")
+	}
+}()
+
+// maxScanKeys is the number of maximum number of keys the seedScanner will
+// generate before giving up.
+var maxScanKeys = func() uint64 {
+	switch build.Release {
+	case "dev":
+		return 1e6
+	case "standard":
+		return 100e6
+	case "testing":
+		return 1e5
+	default:
+		panic("unrecognized build.Release")
+	}
+}()
+
+var errMaxKeys = fmt.Errorf("refused to generate more than %v keys from seed", maxScanKeys)
+
+type scannedOutput struct {
+	id        types.OutputID
+	value     types.Currency
+	seedIndex uint64
+}
+
+// A seedScanner scans the blockchain for addresses that belong to a given
+// seed.
+type seedScanner struct {
+	blockheight      types.BlockHeight
+	dustThreshold    types.Currency              // minimum value of outputs to be included
+	keys             map[types.UnlockHash]uint64 // map address to seed index
```

Comment:
you need the unlock hashes to compare against the addresses that appear in the blockchain. Perhaps you could optimize by truncating the hash and then doing a follow-up full comparison if you get a match. i.e.:
```go
type truncatedHash [8]byte

keys map[truncatedHash]uint64

// check if addr is a seed address
if index, ok := keys[addr]; ok {
    // partial match; check for full match
    fullHash := generateKey(seed, index).UnlockHash()
    if fullHash == addr {
        // full match
    }
}
```